Lib/http/client.py

```diff
@@ -1085,18 +1085,15 @@ def putrequest(self, method, url, skip_host=False,
         else:
             raise CannotSendRequest(self.__state)
 
-        # Save the method we use, we need it later in the response phase
+        # Save the method for use later in the response phase
         self._method = method
-        if not url:
-            url = '/'
-        # Prevent CVE-2019-9740.
-        if match := _contains_disallowed_url_pchar_re.search(url):
-            raise InvalidURL(f"URL can't contain control characters. {url!r} "
-                             f"(found at least {match.group()!r})")
+
+        url = url or '/'
+        self._validate_path(url)
+
         request = '%s %s %s' % (method, url, self._http_vsn_str)
 
-        # Non-ASCII characters should have been eliminated earlier
-        self._output(request.encode('ascii'))
+        self._output(self._encode_request(request))
```

Comment:
overall I think I do like this PR better.  It seems... cleaner.  The one thing of note different here vs #16321 is effectively this call.  encoding and validation are entirely separate, and the encoding applies to the entire http request line including the method and version.  i like it.

Reply:
This approach is really similar to the first draft I'd proposed, but it had these issues.

- it adds overhead of two function calls to the routine (probably not a big deal)
- it requires the overriding tool to possibly override multiple functions
- it doesn't give an overriding tool the ability to accept something other than a string for the url/path.

It was these reasons that led me toward the approach in the other proposal.

But I agree, this implementation, while providing less freedom of customization, is simpler, easier to read, and likely achieves everything that's needed.